packages/expression-evaluator/lib/evaluators/ContextualizedEvaluator.ts

```diff
@@ -0,0 +1,136 @@
+import type { MediatorQueryOperation } from '@comunica/bus-query-operation';
+import { ActorQueryOperation, materializeOperation } from '@comunica/bus-query-operation';
+import type { FunctionBusType, IActionContext } from '@comunica/types';
+import type * as RDF from '@rdfjs/types';
+import { LRUCache } from 'lru-cache';
+import * as E from '../expressions';
+import { expressionToVar } from '../functions/Helpers';
+import type { FunctionArgumentsCache } from '../functions/OverloadTree';
+import { AlgebraTransformer } from '../transformers/AlgebraTransformer';
+import type { ITimeZoneRepresentation } from '../util/DateTimeHelpers';
+import { extractTimeZone } from '../util/DateTimeHelpers';
+import * as Err from '../util/Errors';
+import type { SuperTypeCallback, TypeCache, ISuperTypeProvider } from '../util/TypeHandling';
+
+export type AsyncExtensionFunction = (args: RDF.Term[]) => Promise<RDF.Term>;
+export type AsyncExtensionFunctionCreator = (functionNamedNode: RDF.NamedNode) =>
+Promise<AsyncExtensionFunction | undefined>;
+
+export interface IAsyncEvaluatorContext {
+  now?: Date;
+  baseIRI?: string;
+  typeCache?: TypeCache;
+  getSuperType?: SuperTypeCallback;
+  functionArgumentsCache?: FunctionArgumentsCache;
+  defaultTimeZone?: ITimeZoneRepresentation;
+  actionContext: IActionContext;
+  mediatorQueryOperation: MediatorQueryOperation;
+  mediatorFunction: FunctionBusType;
+}
+
+/**
+ * This class provides evaluation functionality to already transformed expressions.
+ * It also holds all context items needed for evaluating functions.
+ */
+export class ContextualizedEvaluator {
+  public readonly transformer: AlgebraTransformer;
+
+  private readonly subEvaluators: Record<string,
+  (expr: E.Expression, mapping: RDF.Bindings) => Promise<E.Term> | E.Term> =
+      {
+        [E.ExpressionType.Term]: this.term.bind(this),
+        [E.ExpressionType.Variable]: this.variable.bind(this),
+        [E.ExpressionType.Operator]: this.evalFunction.bind(this),
+        [E.ExpressionType.SpecialOperator]: this.evalFunction.bind(this),
+        [E.ExpressionType.Named]: this.evalFunction.bind(this),
+        [E.ExpressionType.Existence]: this.evalExistence.bind(this),
+        [E.ExpressionType.Aggregate]: this.evalAggregate.bind(this),
+      };
+
+  // Context items
+  public readonly extensionFunctionCreator?: AsyncExtensionFunctionCreator;
+  public readonly now: Date;
+  public readonly baseIRI?: string;
+  public readonly functionArgumentsCache: FunctionArgumentsCache;
+  public readonly superTypeProvider: ISuperTypeProvider;
+  public readonly defaultTimeZone: ITimeZoneRepresentation;
+  public readonly actionContext: IActionContext;
+  public readonly mediatorQueryOperation: MediatorQueryOperation;
+  public readonly mediatorFunction: FunctionBusType;
+
+  public constructor(context: IAsyncEvaluatorContext) {
```

Comment:
Since this class essentially materialized the evaluator context, shall we rename it to `MaterializedEvaluatorContext`?